source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -43,49 +46,140 @@ struct CompressorRegistry : public StreamInfo::FilterState::Object {
+              : proto_config.disable_on_etag_header()),
+      remove_accept_encoding_header_(
+          proto_config.has_response_direction_config()
+              ? proto_config.response_direction_config().remove_accept_encoding_header()
+              : proto_config.remove_accept_encoding_header()),
+      response_stats_{generateResponseStats(stats_prefix, scope)} {}
+
+const envoy::extensions::filters::http::compressor::v3::Compressor::CommonDirectionConfig
+CompressorFilterConfig::ResponseDirectionConfig::commonConfig(
+    const envoy::extensions::filters::http::compressor::v3::Compressor& proto_config) {
+  if (proto_config.has_response_direction_config()) {
+    return proto_config.response_direction_config().common_config();
+  }
+  envoy::extensions::filters::http::compressor::v3::Compressor::CommonDirectionConfig config = {};
+  if (proto_config.has_content_length()) {
+    config.set_allocated_content_length(
+        // According to
+        // https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#embeddedmessage
+        // the message Compressor takes ownership of the allocated Protobuf::Uint32Value object.
+        new Protobuf::UInt32Value(proto_config.content_length()));
+  }
+  // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDeleteLeaks)
+  for (const std::string& ctype : proto_config.content_type()) {
+    config.add_content_type(ctype);
+  }
+  config.set_allocated_enabled(
+      // According to
+      // https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#embeddedmessage
+      // the message Compressor takes ownership of the allocated Protobuf::Uint32Value object.
+      new envoy::config::core::v3::RuntimeFeatureFlag(proto_config.runtime_enabled()));
+  // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDeleteLeaks)
+  return config;
+}
+
 CompressorFilter::CompressorFilter(const CompressorFilterConfigSharedPtr config)
-    : skip_compression_{true}, config_(std::move(config)) {}
+    : config_(std::move(config)) {}
 
-Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                          bool end_stream) {
   const Http::HeaderEntry* accept_encoding = headers.getInline(accept_encoding_handle.handle());
   if (accept_encoding != nullptr) {
     // Capture the value of the "Accept-Encoding" request header to use it later when making
     // decision on compressing the corresponding HTTP response.
     accept_encoding_ = std::make_unique<std::string>(accept_encoding->value().getStringView());
   }
 
-  if (config_->enabled() && config_->removeAcceptEncodingHeader()) {
+  const auto& response_config = config_->responseDirectionConfig();
+  if (response_config.compressionEnabled() && response_config.removeAcceptEncodingHeader()) {
     headers.removeInline(accept_encoding_handle.handle());
```

Comment:
AFAIU this option is needed prevent pre-mature response compressing at upstream hops, so that 1) the downstream (a client or a caching filter) will always receive a stream compressed with well known parameters or 2) the stream is always compressed with a hardware accelerator. In these scenarios there's no need to disable compression unconditionally in the filter and in any upstream hops at the same time. I guess a cleaner approach would be to modify headers in a `RouteConfiguration`.